src/sinks/util/service/health.rs

```diff
@@ -0,0 +1,188 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+use futures::{ready, FutureExt};
+use futures_util::{future::BoxFuture, TryFuture};
+use pin_project::pin_project;
+use tokio::time::{sleep, Duration};
+use tower::{load::Load, Service};
+
+use crate::{
+    emit,
+    internal_events::{EndpointsActive, OpenGauge, OpenToken},
+    sinks::util::retries::RetryLogic,
+};
+
+enum ServiceState {
+    Healthcheck(BoxFuture<'static, bool>),
+    Backoff {
+        timer: BoxFuture<'static, ()>,
+        healthcheck: bool,
+    },
+    Ready(OpenToken<fn(usize)>),
+}
+
+/// A service which estimates load on inner service and
+/// also forces backoff if the inner service is not healthy.
+pub struct HealthService<S, RL> {
+    inner: S,
+    healthcheck: Box<dyn Fn() -> BoxFuture<'static, bool> + Send>,
+    logic: RL,
+    reactivate_delay: Duration,
+    /// Serves as an estimate of load and for notifying about errors.
+    /// 0 - everything is fine
+    /// 1 - check health
+    /// 2 - backoff
+    request_handle: Arc<AtomicUsize>,
```

Comment:
I'd like to see this avoid magic constants for an enum. Could this be accomplished by using [`AtomicCell` in crossbeam](https://docs.rs/crossbeam-utils/latest/crossbeam_utils/atomic/struct.AtomicCell.html) to this effect?